### PR TITLE
Cw655 add int

### PR DESCRIPTION
**Pull request by Janicewei**

### CODE REVIEW

**Comment by vb000** [comment]
Is this working in cosim? I'm seeing these errors:

```
        Offending '(~id_r.decode.unsupported)'
Error: "/mnt/users/ssd1/no_backup/bandhav/bsg_bladerunner/bsg_manycore/v/vanilla_bean/vanilla_core.v", 1599: tb.card.fpga.CL.manycore_wrapper.manycore.y[3].x[1].tile.proc.h.z.vcore.unnamed$$_2: at time 46625600000 ps
Unsupported instruction: 0279bb33
```

**Comment by cbatten** [comment]
@Janicewei have we tested this on cosim?

**Comment by Janicewei** [comment]
Just tested again. It's working on cosim.

**Comment by vb000** [comment]
Might be some problem with my setup.

**Comment by vb000** [comment]
> Just tested again. It's working on cosim.

Could you try running pytest with `-vs` falgs? The tests are passing, in spite of these error messages...

**Comment by Janicewei** [comment]
> Could you try running pytest with `-vs` falgs? The tests are passing, in spite of these error messages...

Should I try 
pytest -vs pytest_runner.py?

**Comment by vb000** [comment]
> > Could you try running pytest with `-vs` falgs? The tests are passing, in spite of these error messages...
> 
> Should I try
> pytest -vs pytest_runner.py?

Just the add_sub test should be fine... `pycosim -m pytest test_add_sub.py -vs` in hammerblade/torch/tests

**Comment by Janicewei** [comment]
> Just the add_sub test should be fine... `pycosim -m pytest test_add_sub.py -vs` in hammerblade/torch/tests

Is it supposed to take a while? The test is still running.

**Comment by vb000** [comment]
Oh sorry, just a normal 64-bit test produced this message... no need to run hypothesis test:
`pycosim -m pytest test_add_sub.py -vs -k-hypothesis`

**Comment by Janicewei** [comment]
> Oh sorry, just a normal 64-bit test produced this message... no need to run hypothesis test:
> `pycosim -m pytest test_add_sub.py -vs -k-hypothesis`

```
test_add_sub.py::test_add_1 "/work/global/brg/install/bare-pkgs/x86_64-centos7/pkgs/bsg_bladerunner_v4.0.0b/bsg_manycore/v/bsg_manycore_endpoint_standard.v", 346: tb.card.fpga.CL.mcl_to_axil.mc_ep_to_fifos.epsd.genblk2.unnamed$$_0: started at 33328000ps failed at 33328000ps
        Offending '((out_credits_o === 'x) || (out_credits_o > 5'b0))'
## out of remote store credits(= 0) x,y=0,1 displaying only once (tb.card.fpga.CL.mcl_to_axil.mc_ep_to_fifos.epsd)
##   (this may be a performance problem; or normal behavior)
```
I didn't get the message you mentioned but got above before unfreezing tiles

**Comment by yodada** [comment]
Emm one thing I want to point out is, I think Bandhav is on v4.1.0 while Janice still uses the old pipeline. That probably explains the difference in behavior

**Comment by yodada** [comment]
So what is the current status on this? Does it fail on COSIM?

**Comment by yodada** [comment]
FYI we are investigating the unsupported instruction issue

**Comment by drichmond** [comment]
> > Oh sorry, just a normal 64-bit test produced this message... no need to run hypothesis test:
> > `pycosim -m pytest test_add_sub.py -vs -k-hypothesis`
> 
> ```
> test_add_sub.py::test_add_1 "/work/global/brg/install/bare-pkgs/x86_64-centos7/pkgs/bsg_bladerunner_v4.0.0b/bsg_manycore/v/bsg_manycore_endpoint_standard.v", 346: tb.card.fpga.CL.mcl_to_axil.mc_ep_to_fifos.epsd.genblk2.unnamed$$_0: started at 33328000ps failed at 33328000ps
>         Offending '((out_credits_o === 'x) || (out_credits_o > 5'b0))'
> ## out of remote store credits(= 0) x,y=0,1 displaying only once (tb.card.fpga.CL.mcl_to_axil.mc_ep_to_fifos.epsd)
> ##   (this may be a performance problem; or normal behavior)
> ```
> 
> I didn't get the message you mentioned but got above before unfreezing tiles

This is not an issue. This is just an artifact of flooding the network. See https://github.com/bespoke-silicon-group/bsg_replicant/issues/449

**Comment by yodada** [comment]
Yeah I understand that warning is fine.
What I meant is this one
```
Error: "/work/global/brg/install/bare-pkgs/x86_64-centos7/pkgs/bsg_bladerunner_v4.1.0/bsg_manycore/v/vanilla_bean/vanilla_core.v", 1599: tb.card.fpga.CL.network.manycore_wrapper.manycore.y[2].x[3].tile.proc.h.z.vcore.unnamed$$_2: at time 122172000000 ps
Unsupported instruction: 02b6b6b3
```

**Comment by drichmond** [comment]
yep! Understood. I was just clearing up a question that was asked earlier

**Comment by yodada** [comment]
This is the offending instruction
```
0279bb33            mulhu x22,x19,x7
```
emmm ... so `mulhu` is not supported?

**Comment by yodada** [comment]
It looks like `mulhu` is the problem. All unsupported instruction reported above have opcode `7b'0110011`

**Comment by vb000** [comment]
> So what is the current status on this? Does it fail on COSIM?

Sorry I had totally missed this! I had observed on an untagged version (I was working on all masters of bsg_bladerunner). So, v4.1.0 would most probably have this issue. Then I removed int64 kernel call from batch norm because it required full kernel launch for just one in64, int64 add! Yes, I observed this on cosim but it didn't fail for some reason. The test still passed.

**Comment by yodada** [comment]
not sure what we used in v4.0.0 ... there must be a compiler flag that enabled emitting this instruction

**Comment by vb000** [comment]
Yeah `mulhu` is unsupported...

**Comment by yodada** [comment]
we had a brief discussion, and it seems to us there is no simple way to emulate `mulh` other than shift-add ... is that true?

File: aten/src/ATen/Dispatch.h [hunk 1]
```diff
@@ -108,6 +108,20 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                        \
   }()
 
+#define AT_DISPATCH_FLOAT_AND_LONG(TYPE, NAME, ...)                         \
```
Comment (yodada) [inline aten/src/ATen/Dispatch.h]: Since we also support int32_t here in the marco, maybe we should name it `AT_DISPATCH_FLOAT_AND_INTS`?

File: c10/hammerblade/CMakeLists.txt [hunk 2]
```diff
@@ -120,11 +120,8 @@ endforeach()
 install(FILES ${CMAKE_BINARY_DIR}/c10/hammerblade/impl/hammerblade_cmake_macros.h
   DESTINATION include/c10/hammerblade/impl)
 
-file(GLOB KERNEL_FILES ${CMAKE_CURRENT_BINARY_DIR}/riscv/kernel.riscv
-                       ${CMAKE_CURRENT_BINARY_DIR}/riscv/*.dis)
-
 if(NOT USE_HB_EMUL)
-  install(FILES ${KERNEL_FILES} DESTINATION riscv)
+  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/kernel.riscv DESTINATION riscv)
```
Comment (yodada) [inline c10/hammerblade/CMakeLists.txt]: I'll take care of this. Don't worry

File: hammerblade/torch/kernel/kernel_add.cpp [hunk 3]
```diff
@@ -1,39 +1,152 @@
...
+  
+  if (total_l - alhb_l != al * bl) {
+    carry = 1;
+  }
+  
+  int32_t _mulh = ahbh + alhb_h + carry;
+  return _mulh;
+}
+
+inline int64_t mulh(int64_t a, int64_t b) {
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_add.cpp]: ahhhh you dont need to have a int64_t version of the mulh ;)

File: hammerblade/torch/kernel/kernel_add.cpp [hunk 4]
```diff
@@ -1,39 +1,152 @@
...
+  HBTensor<long long> c(t0_p);
+  HBTensor<long long> a(t1_p);
+  HBTensor<long long> b(t2_p);
+  long long alpha = *alpha_p;
 
+  bsg_cuda_print_stat_kernel_start();
+
+  hb_tiled_foreach(
+    [alpha](long long a, long long b) {
+      return a + alpha * b;
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_add.cpp]: this is where you will be using that mulh ;)

**Comment by yodada** [comment]
Finally we emulated mulh with just mul. The math was easy, writing the code right was not ...
https://github.com/cornell-brg/hb-pytorch/blob/7b3b0a816303d097244870185baecc07568526f8/hammerblade/torch/kernel/kernel_add.cpp#L14-L44

File: hammerblade/torch/kernel/kernel_add.cpp [hunk 5]
```diff
@@ -1,39 +1,149 @@
...
+  bsg_cuda_print_stat_kernel_start();
+
+  hb_tiled_foreach(
+    [alpha](long long a, long long b) {
+      uint32_t ah = alpha >> 32;
+      uint32_t al = alpha & 0xffffffff;
+      uint32_t bh = b >> 32;
+      uint32_t bl = b & 0xffffffff;
+      uint32_t albl = al * bl;
+      int64_t ab2 = ((int64_t)(ah * bl + al * bh) << 32) + (int64_t)(((uint64_t)mulh(al,bl)) << 32) + albl;
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_add.cpp]: Maybe we can give it a more meaningful name? I used `ab2` in my test but it's not a good one for "production"

Comment (Janicewei) [reply hammerblade/torch/kernel/kernel_add.cpp]: sonds good! changed it to `mul`

**Review by yodada — APPROVED** [review]
Awesome!
